File: patterns/01-design-patterns/02-structural/01-adapter/go/adapter_test.go
package adapter

import (
	"context"
	"encoding/json"
	"errors"
	"os"
	"testing"
)

type contractCase struct {
	ID       string `json:"id"`
	Category string `json:"category"`
	Input    struct {
		Request struct {
			Model         string    `json:"model"`
			Messages      []Message `json:"messages"`
			RequiresTools bool      `json:"requires_tools"`
		} `json:"request"`
		ProviderResponse *struct {
			Output   string `json:"output"`
			StopCode string `json:"stop_code"`
		} `json:"provider_response"`
		ProviderError *ProviderError `json:"provider_error"`
	} `json:"input"`
	Expected struct {
		MappedRequest *LegacyRequest `json:"mapped_request"`
		Response      *struct {
			Content      string `json:"content"`
			FinishReason string `json:"finish_reason"`
		} `json:"response"`
		ProviderCalls int `json:"provider_calls"`
	} `json:"expected"`
	ExpectedError *struct {
		Code      string `json:"code"`
		Retryable bool   `json:"retryable"`
	} `json:"expected_error"`
}

type contractFixture struct {
	PatternID string         `json:"pattern_id"`
	Cases     []contractCase `json:"cases"`
}

type fakeLegacyClient struct {
	response    LegacyResponse
	err         error
	lastRequest *LegacyRequest
	calls       int
}

func (f *fakeLegacyClient) Generate(_ context.Context, request LegacyRequest) (LegacyResponse, error) {
	f.calls++
	f.lastRequest = &request
	return f.response, f.err
}

func loadFixture(t *testing.T) contractFixture {
	t.Helper()
	body, err := os.ReadFile("../fixtures/contract.json")
	if err != nil {
		t.Fatal(err)
	}
	var fixture contractFixture
	if err := json.Unmarshal(body, &fixture); err != nil {
		t.Fatal(err)
	}
	return fixture
}

func findCase(t *testing.T, fixture contractFixture, id string) contractCase {
	t.Helper()
	for _, candidate := range fixture.Cases {
		if candidate.ID == id {
			return candidate
		}
	}
	t.Fatalf("missing contract case %q", id)
	return contractCase{}
}

func requestFrom(testCase contractCase) ChatRequest {
	return ChatRequest{
		Model:         testCase.Input.Request.Model,
		Messages:      testCase.Input.Request.Messages,
		RequiresTools: testCase.Input.Request.RequiresTools,
	}
}

func TestMapsSharedContractFixture(t *testing.T) {
	testCase := findCase(t, loadFixture(t), "maps_request_and_response")
	provider := testCase.Input.ProviderResponse
	client := &fakeLegacyClient{response: LegacyResponse{Output: provider.Output, StopCode: provider.StopCode}}
	adapter := NewLegacyProviderAdapter(client)

	response, err := adapter.Complete(context.Background(), requestFrom(testCase))
	if err != nil {
		t.Fatal(err)
	}

	if *client.lastRequest != *testCase.Expected.MappedRequest {
		t.Fatalf("mapped request = %#v, want %#v", *client.lastRequest, *testCase.Expected.MappedRequest)
	}
	if response.Content != testCase.Expected.Response.Content || response.FinishReason != testCase.Expected.Response.FinishReason {
		t.Fatalf("response = %#v, want %#v", response, *testCase.Expected.Response)
	}
}

func TestFailsExplicitlyForUnknownModel(t *testing.T) {
	testCase := findCase(t, loadFixture(t), "rejects_unknown_model")
	_, err := NewLegacyProviderAdapter(&fakeLegacyClient{}).Complete(context.Background(), requestFrom(testCase))
	assertNormalizedError(t, err, testCase.ExpectedError.Code, testCase.ExpectedError.Retryable)
}

func TestNormalizesProviderError(t *testing.T) {
	testCase := findCase(t, loadFixture(t), "normalizes_provider_error")
	client := &fakeLegacyClient{err: testCase.Input.ProviderError}
	_, err := NewLegacyProviderAdapter(client).Complete(context.Background(), requestFrom(testCase))
	assertNormalizedError(t, err, testCase.ExpectedError.Code, testCase.ExpectedError.Retryable)
}

func TestRejectsUnsupportedBeforeProviderCall(t *testing.T) {
	testCase := findCase(t, loadFixture(t), "rejects_unsupported_before_provider_call")
	client := &fakeLegacyClient{}
	_, err := NewLegacyProviderAdapter(client).Complete(context.Background(), requestFrom(testCase))
	assertNormalizedError(t, err, testCase.ExpectedError.Code, testCase.ExpectedError.Retryable)
	if client.calls != testCase.Expected.ProviderCalls {
		t.Fatalf("provider calls = %d, want %d", client.calls, testCase.Expected.ProviderCalls)
	}
}

func assertNormalizedError(t *testing.T, err error, code string, retryable bool) {
	t.Helper()
	var normalized *NormalizedError
	if !errors.As(err, &normalized) {
		t.Fatalf("error = %T, want *NormalizedError", err)
	}
	if normalized.Code != code || normalized.Retryable != retryable {
		t.Fatalf("normalized error = %#v, want code=%q retryable=%v", normalized, code, retryable)
	}
}

File: patterns/01-design-patterns/02-structural/01-adapter/python/adapter_test.py
import json
from pathlib import Path

import pytest

from adapter import (
    ChatRequest,
    LegacyProviderAdapter,
    LegacyResponse,
    Message,
    NormalizedError,
    ProviderError,
)


CONTRACT = json.loads(
    (Path(__file__).parents[1] / "fixtures" / "contract.json").read_text()
)


def contract_case(case_id):
    return next(case for case in CONTRACT["cases"] if case["id"] == case_id)


class FakeLegacyClient:
    def __init__(self, response=None, error=None):
        self.response = response
        self.error = error
        self.last_request = None
        self.calls = 0

    def generate(self, request):
        self.calls += 1
        self.last_request = request
        if self.error is not None:
            raise self.error
        return self.response


def request_from(case):
    request = case["input"]["request"]
    return ChatRequest(
        model=request["model"],
        messages=tuple(Message(**message) for message in request["messages"]),
        requires_tools=request["requires_tools"],
    )


def test_maps_shared_contract_fixture():
    case = contract_case("maps_request_and_response")
    provider = case["input"]["provider_response"]
    client = FakeLegacyClient(
        response=LegacyResponse(output=provider["output"], stop_code=provider["stop_code"])
    )

    response = LegacyProviderAdapter(client).complete(request_from(case))

    assert vars(client.last_request) == case["expected"]["mapped_request"]
    assert response.content == case["expected"]["response"]["content"]
    assert response.finish_reason == case["expected"]["response"]["finish_reason"]


def test_fails_explicitly_for_unknown_model():
    case = contract_case("rejects_unknown_model")

    with pytest.raises(NormalizedError) as captured:
        LegacyProviderAdapter(FakeLegacyClient()).complete(request_from(case))

    assert captured.value.code == case["expected_error"]["code"]
    assert captured.value.retryable is case["expected_error"]["retryable"]


def test_normalizes_provider_error():
    case = contract_case("normalizes_provider_error")
    provider_error = case["input"]["provider_error"]
    client = FakeLegacyClient(
        error=ProviderError(provider_error["code"], provider_error["message"])
    )

    with pytest.raises(NormalizedError) as captured:
        LegacyProviderAdapter(client).complete(request_from(case))

    assert captured.value.code == case["expected_error"]["code"]
    assert captured.value.retryable is case["expected_error"]["retryable"]


def test_rejects_unsupported_before_provider_call():
    case = contract_case("rejects_unsupported_before_provider_call")
    client = FakeLegacyClient()

    with pytest.raises(NormalizedError) as captured:
        LegacyProviderAdapter(client).complete(request_from(case))

    assert captured.value.code == case["expected_error"]["code"]
    assert captured.value.retryable is case["expected_error"]["retryable"]
    assert client.calls == case["expected"]["provider_calls"]

